File: Codes/2786-visit-array-positions-to-maximize-score.cpp
/*
 * @Author: LetMeFly
 * @Date: 2024-06-14 16:40:33
 * @LastEditors: LetMeFly
 * @LastEditTime: 2024-06-14 16:45:41
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

typedef long long ll;
class Solution {
public:
    ll maxScore(vector<int>& nums, int x) {
        ll odd = nums[0] % 2 ? 0 : -x, even = nums[0] % 2 ? -x : 0;
        for (int t : nums) {
            if (t % 2) {
                odd = max(odd + t, even + t - x);
            }
            else {
                even = max(odd + t - x, even + t);
            }
        }
        return max(odd, even);
    }
};

File: Codes/2786-visit-array-positions-to-maximize-score.py
'''
Author: LetMeFly
Date: 2024-06-14 19:01:46
LastEditors: LetMeFly
LastEditTime: 2024-06-14 19:03:16
'''
from typing import List

class Solution:
    def maxScore(self, nums: List[int], x: int) -> int:
        odd, even = 0 if nums[0] % 2 else -x, -x if nums[0] % 2 else 0
        for t in nums:
            if t % 2:
                odd = max(odd + t, even + t - x)
            else:
                even = max(odd + t - x, even + t)
        return max(even, odd)